org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/TryWithResourcesFilter.java

```diff
@@ -0,0 +1,426 @@
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis.filter;
+
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.JumpInsnNode;
+import org.objectweb.asm.tree.LabelNode;
+import org.objectweb.asm.tree.MethodInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.TryCatchBlockNode;
+import org.objectweb.asm.tree.VarInsnNode;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Filters code that is generated for try-with-resources statement.
+ */
+public final class TryWithResourcesFilter implements IFilter {
+
+	public void filter(final MethodNode methodNode,
+			final IFilterOutput output) {
+		if (methodNode.tryCatchBlocks.isEmpty()) {
+			return;
+		}
+		final Matcher matcher = new Matcher(output);
+		for (TryCatchBlockNode t : methodNode.tryCatchBlocks) {
+			if (t.type == null) {
+				matcher.start(t.handler);
+				if (!matcher.matchEcj()) {
+					matcher.start(t.handler);
+					matcher.matchEcjNoFlowOut();
+				}
+			} else if ("java/lang/Throwable".equals(t.type)) {
+				for (Matcher.JavacPattern p : Matcher.JavacPattern.values()) {
+					matcher.start(t.handler);
+					if (matcher.matchJavac(p)) {
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	static class Matcher {
+		private final IFilterOutput output;
+
+		private final Map<String, VarInsnNode> vars = new HashMap<String, VarInsnNode>();
+		private final Map<String, String> owners = new HashMap<String, String>();
+		private final Map<String, LabelNode> labels = new HashMap<String, LabelNode>();
+
+		private AbstractInsnNode start;
+		private AbstractInsnNode cursor;
+
+		Matcher(final IFilterOutput output) {
+			this.output = output;
+		}
+
+		private enum JavacPattern {
+			OPTIMAL, FULL, OMITTED_NULL_CHECK, METHOD,
+		}
+
+		private void start(final AbstractInsnNode start) {
+			this.start = start;
+			cursor = start.getPrevious();
+			vars.clear();
+			labels.clear();
+			owners.clear();
+		}
+
+		private boolean matchJavac(final JavacPattern p) {
+			// "catch (Throwable t)"
+			if (!nextIsVar(Opcodes.ASTORE, "t1")) {
+				return false;
+			}
+			// "primaryExc = t"
+			if (!nextIsVar(Opcodes.ALOAD, "t1")) {
+				return false;
+			}
+			if (!nextIsVar(Opcodes.ASTORE, "primaryExc")) {
+				return false;
+			}
+			// "throw t"
+			if (!nextIsVar(Opcodes.ALOAD, "t1")) {
+				return false;
+			}
+			if (!nextIs(Opcodes.ATHROW)) {
+				return false;
+			}
+
+			// "catch (any t)"
+			if (!nextIsVar(Opcodes.ASTORE, "t2")) {
+				return false;
+			}
+			if (!nextIsJavacClose(p)) {
+				return false;
+			}
+			// "throw t"
+			if (!nextIsVar(Opcodes.ALOAD, "t2")) {
+				return false;
+			}
+			if (!nextIs(Opcodes.ATHROW)) {
+				return false;
+			}
+
+			final AbstractInsnNode end = cursor;
+			AbstractInsnNode c = start.getPrevious();
```

Comment:
@Godin Looks like local variable c is not just a local loop counter but also carries some semantics here (as used later in `ignore()`. Maybe a name could explain what it is meant for (`startOfXxx`).